server/src/main/java/org/opensearch/node/remotestore/RemoteStoreNodeAttribute.java

```diff
@@ -144,62 +169,117 @@ private RepositoryMetadata buildRepositoryMetadata(DiscoveryNode node, String na
+            }
         }
-        if (node.getAttributes().containsKey(REMOTE_STORE_ROUTING_TABLE_REPOSITORY_NAME_ATTRIBUTE_KEY)) {
-            repositoryNames.add(validateAttributeNonNull(node, REMOTE_STORE_ROUTING_TABLE_REPOSITORY_NAME_ATTRIBUTE_KEY));
+        return null;
+    }
+
+    private Map<String, String> getValidatedRepositoryNames(DiscoveryNode node) {
+        Set<Tuple<String, String>> repositoryNames = new HashSet<>();
+        if (containsKey(node.getAttributes(), REMOTE_SEGMENT_REPOSITORY_NAME_ATTRIBUTE_KEYS)
+            || containsKey(node.getAttributes(), REMOTE_TRANSLOG_REPOSITORY_NAME_ATTRIBUTE_KEYS)) {
+            repositoryNames.add(validateAttributeNonNull(node, REMOTE_SEGMENT_REPOSITORY_NAME_ATTRIBUTE_KEYS));
+            repositoryNames.add(validateAttributeNonNull(node, REMOTE_TRANSLOG_REPOSITORY_NAME_ATTRIBUTE_KEYS));
+            repositoryNames.add(validateAttributeNonNull(node, REMOTE_CLUSTER_STATE_REPOSITORY_NAME_ATTRIBUTE_KEYS));
+        } else if (containsKey(node.getAttributes(), REMOTE_CLUSTER_STATE_REPOSITORY_NAME_ATTRIBUTE_KEYS)) {
+            repositoryNames.add(validateAttributeNonNull(node, REMOTE_CLUSTER_STATE_REPOSITORY_NAME_ATTRIBUTE_KEYS));
         }
+        if (containsKey(node.getAttributes(), REMOTE_ROUTING_TABLE_REPOSITORY_NAME_ATTRIBUTE_KEYS)) {
+            repositoryNames.add(validateAttributeNonNull(node, REMOTE_ROUTING_TABLE_REPOSITORY_NAME_ATTRIBUTE_KEYS));
+        }
+
+        Map<String, String> repoNamesWithPrefix = new HashMap<>();
+        repositoryNames.forEach(t -> {
+            String[] attrKeyParts = t.v2().split("\\.");
+            repoNamesWithPrefix.put(t.v1(), attrKeyParts[0]);
+        });
 
-        return repositoryNames;
+        return repoNamesWithPrefix;
     }
 
     public static boolean isRemoteStoreAttributePresent(Settings settings) {
-        return settings.getByPrefix(Node.NODE_ATTRIBUTES.getKey() + REMOTE_STORE_NODE_ATTRIBUTE_KEY_PREFIX).isEmpty() == false;
+        for (String prefix : REMOTE_STORE_NODE_ATTRIBUTE_KEY_PREFIX) {
+            if (settings.getByPrefix(Node.NODE_ATTRIBUTES.getKey() + prefix).isEmpty() == false) {
+                return true;
+            }
+        }
+        return false;
     }
 
     public static boolean isRemoteDataAttributePresent(Settings settings) {
-        return settings.getByPrefix(Node.NODE_ATTRIBUTES.getKey() + REMOTE_STORE_SEGMENT_REPOSITORY_NAME_ATTRIBUTE_KEY).isEmpty() == false
-            || settings.getByPrefix(Node.NODE_ATTRIBUTES.getKey() + REMOTE_STORE_TRANSLOG_REPOSITORY_NAME_ATTRIBUTE_KEY).isEmpty() == false;
+        return isSegmentRepoConfigured(settings) || isTranslogRepoConfigured(settings);
+    }
+
+    public static boolean isSegmentRepoConfigured(Settings settings) {
+        for (String prefix : REMOTE_SEGMENT_REPOSITORY_NAME_ATTRIBUTE_KEYS) {
+            if (settings.getByPrefix(Node.NODE_ATTRIBUTES.getKey() + prefix).isEmpty() == false) {
+                return true;
+            }
```

Comment:
the logic essentially for all the prefix is which ever repo is present will be returned? 
I am trying to create mental model as user, when i should configure what? why there has to be list of prefix?

Reply:
Yes the repo will be returned based on Any prefix match. The user might be required to configure either `remote_store`.*  (or) `remote_publication.*` only when cluster is migrating from remote datastore to publication. I will not expose this setting to user and keep it internal for now.